enhancements/update/upgrades-blocking-on-ack.md

```diff
@@ -0,0 +1,210 @@
+This enhancement adds a generic upgrade blocking mechanism for blocking minor level upgrades.
+This generic upgrade blocker can consist of one or more items each of which will block a minor level upgrade.
+Each generic upgrade blocker item will be associated with some well-defined task to be completed by an administrator.
+Once the task has been completed the administrator will acknowledge completion and its associated item will no longer block minor level upgrades.
+
+## Motivation
+
+OCP 4.9 uses kubernetes v1.22 which is removing a large number of [APIs][k8s-1.22-removed-apis].
+We suspect that many customer clusters are still using these APIs through helm charts, operators, generic workloads, administrative tools, etc.
+We believe there is no effective way for a Cluster Operator to detect the use of these APIs by external applications.
+Relying solely on automated API detection to block upgrades could result in blocking for APIs that are not actively used any longer or not blocking because an application was not active during the check.
+Having a human in the loop to drive the analysis and make the final decision is the surest way to prevent customers from upgrading from 4.8.z to 4.9.z and encountering missing API related breakage.
+
+### Goals
+
+* Acknowledgment requires the explicit update of a cluster resource by an administrator.
+* For the initial release in 4.8.z, administrator acknowledgment of API removal will be required before a minor level upgrade to 4.9 is allowed.
+* Upgrade blocking mechanism must be generic such that it can be repurposed in the future.
+* It must be clear what an administrator is acknowledging.
+* Administrators can bypass the upgrade blocking mechanism by using the existing CVO "force" implementation.
+* The UI will clearly indicate when this upgrade blocking mechanism is blocking upgrades, why, and what needs to be done to clear the block.
+* This upgrade blocking mechanism must be compatible with SD, ACM, IKS/ROKS, Hive, and Hypershift.
+
+### Non-Goals
+
+## Proposal
+
+### General Implementation
+
+CVO's existing framework for blocking minor level upgrades will be leveraged.
+The existing framework periodically checks upgradeable conditions.
+A new [upgradeable check method][upgradeableCheck-interface] will be added to the existing framework to implement this upgrade blocking mechanism.
+Use of this existing framework also allows this new upgrade blocking mechanism to be bypassed by using the `force` upgrade option.
+
+The new method's logic will be driven by configmap keys each of which represents a potential minor level upgrade blocker.
+CVO will consume a new configmap, `admin-gates`, defining the keys. For example:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: admin-gates
+  namespace: openshift-cluster-version
+data:
+  ack-4.8-kube-122-api-removals-in-4.9: |
+     Detailed description of what admin needs to do...Read documentation here https://
+  ack-4.8-gate-2: |
```

Comment:
CVO's current ConfigMap handler [uses `mergeMap` for `data`][1], and [`mergeMap` ensures that manifest-side keys have the requested values][2], but does not remove keys which are not mentioned in the manifest.  Are we comfortable with this ConfigMap being append-forever?  Or do we want to define a pruning strategy?  Like "CVO will special-case logic to remove keys from both ConfigMaps unless they match 4.y for the current or a future y"?  Or something?

[1]: https://github.com/openshift/cluster-version-operator/blob/3a68652568e9075c23f491bc8c037942bd67ec82/lib/resourcemerge/core.go#L18
[2]: https://github.com/openshift/cluster-version-operator/blob/3a68652568e9075c23f491bc8c037942bd67ec82/lib/resourcemerge/meta.go#L28-L41

Reply:
personally i don't see an issue w/ append forever, but i'm interested to hear concerns

1) this should be rare
2) we are planning a UX around this(CVO status reporting) so it's not like an admin will need to look through this configmap to figure out what applies to their cluster currently

Reply:
I don't see an issue with it either.